src/serialization/serialization.rs

```diff
@@ -0,0 +1,409 @@
+extern crate byteorder;
+
+use super::{Counter, Histogram};
+use std::io::{self, Cursor, Read, Write};
+use std;
+use super::num::ToPrimitive;
+use self::byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+
+#[path = "tests.rs"]
+#[cfg(test)]
+mod tests;
+
+#[path = "benchmarks.rs"]
+#[cfg(all(test, feature = "bench_private"))]
+mod benchmarks;
+
+const V2_COOKIE_BASE: u32 = 0x1c849303;
+
+const V2_COOKIE: u32 = V2_COOKIE_BASE | 0x10;
+
+const V2_HEADER_SIZE: usize = 40;
+
+/// Errors that occur during serialization.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum SerializeError {
+    /// A count above i64::max_value() cannot be zig-zag encoded, and therefore cannot be
+    /// serialized.
+    CountNotSerializable,
+    /// Internal calculations cannot be represented in `usize`. Use smaller histograms or beefier
+    /// hardware.
+    UsizeTypeTooSmall,
+    /// An i/o operation failed.
+    IoError
+}
+
+impl std::convert::From<std::io::Error> for SerializeError {
+    fn from(_: std::io::Error) -> Self {
+        SerializeError::IoError
+    }
+}
+
+/// Serializer for the V2 binary format.
+/// Serializers are intended to be re-used for many histograms.
+pub struct V2Serializer {
+    buf: Vec<u8>
+}
+
+impl V2Serializer {
+    /// Create a new serializer.
+    pub fn new() -> V2Serializer {
+        V2Serializer {
+            buf: Vec::new()
+        }
+    }
+
+    /// Serialize the histogram.
+    /// Returns the number of bytes written, or an io error.
+    pub fn serialize<T: Counter, W: Write>(&mut self, h: &Histogram<T>, writer: &mut W)
+                                           -> Result<usize, SerializeError> {
+        self.buf.clear();
+        let max_size = Self::max_encoded_size(h).ok_or(SerializeError::UsizeTypeTooSmall)?;
+        self.buf.reserve(max_size);
+
+        self.buf.write_u32::<BigEndian>(V2_COOKIE)?;
+        // placeholder for length
+        self.buf.write_u32::<BigEndian>(0)?;
+        // normalizing index offset
+        self.buf.write_u32::<BigEndian>(0)?;
+        self.buf.write_u32::<BigEndian>(h.significant_value_digits as u32)?;
+        self.buf.write_u64::<BigEndian>(h.lowest_discernible_value)?;
+        self.buf.write_u64::<BigEndian>(h.highest_trackable_value)?;
+        // int to double conversion
+        self.buf.write_f64::<BigEndian>(1.0)?;
+
+        debug_assert_eq!(V2_HEADER_SIZE, self.buf.len());
+
+        unsafe {
+            // want to treat the rest of the vec as a slice, and we've already reserved this
+            // space, so this way we don't have to resize() on a lot of dummy bytes.
+            self.buf.set_len(max_size);
+        }
+
+        let counts_len = encode_counts(h, &mut self.buf[V2_HEADER_SIZE..])?;
+        let total_len = V2_HEADER_SIZE + counts_len;
+
+        // TODO benchmark fastest buffer management scheme
```

Comment:
I'm not yet ready to give up on serialization performance, so this is just a note to future us.